src/draw/lv_draw_dither.c

```diff
@@ -0,0 +1,213 @@
+       The algorithm below is based on few assumptions:
+         1. An error diffusion algorithm (like Floyd Steinberg) here would be hard to implement since it means that a pixel on column n depends on the pixel on row n
+         2. Instead an ordered dithering algorithm shift the value a bit, but the influence only spread from the matrix size (used 8x8 here)
+         3. It means that a pixel i,j only depends on the value of a pixel i-7, j-7 to i,j and no other one.
+       Then we compute a complete row of ordered dither and store it in out. */
+
+    /*The apply the algorithm for this patch*/
+    for(lv_coord_t j = 0; j < w; j++) {
+        int8_t factor = dither_ordered_threshold_matrix[(y & 7) * 8 + ((j) & 7)] - 32;
+        lv_color32_t tmp = grad->hmap[LV_CLAMP(0, j - 4, grad->hmap_size)];
+        lv_color32_t t;
+        t.ch.red   = LV_CLAMP(0, tmp.ch.red + factor, 255);
+        t.ch.green = LV_CLAMP(0, tmp.ch.green + factor, 255);
+        t.ch.blue  = LV_CLAMP(0, tmp.ch.blue + factor, 255);
+
+        grad->map[j] = lv_color_hex(t.full);
+    }
+}
+LV_ATTRIBUTE_FAST_MEM void lv_dither_ordered_ver(lv_gradient_cache_t * grad, lv_coord_t x, lv_coord_t y, lv_coord_t w)
+{
+    /* For vertical dithering, the error is spread on the next column (and not next line).
+       Since the renderer is scanline based, it's not obvious what could be used to perform the rendering efficiently.
+       The algorithm below is based on few assumptions:
+         1. An error diffusion algorithm (like Floyd Steinberg) here would be hard to implement since it means that a pixel on column n depends on the pixel on row n
+         2. Instead an ordered dithering algorithm shift the value a bit, but the influence only spread from the matrix size (used 8x8 here)
+         3. It means that a pixel i,j only depends on the value of a pixel i-7, j-7 to i,j and no other one.
+       Then we compute a complete row of ordered dither and store it in out. */
+
+    /*Extract patch for working with, selected pseudo randomly*/
+    lv_color32_t tmp = grad->hmap[LV_CLAMP(0, y - 4, grad->hmap_size)];
+
+    /*The apply the algorithm for this patch*/
+    for(lv_coord_t j = 0; j < 8; j++) {
+        int8_t factor = dither_ordered_threshold_matrix[(y & 7) * 8 + ((j + x) & 7)] - 32;
+        lv_color32_t t;
+        t.ch.red   = LV_CLAMP(0, tmp.ch.red + factor, 255);
+        t.ch.green = LV_CLAMP(0, tmp.ch.green + factor, 255);
+        t.ch.blue  = LV_CLAMP(0, tmp.ch.blue + factor, 255);
+
+        grad->map[j] = lv_color_hex(t.full);
+    }
+    /*Finally fill the line*/
+    lv_coord_t j = 8;
+    for(; j < w - 8; j += 8) {
+        lv_memcpy(grad->map + j, grad->map, 8 * sizeof(*grad->map));
+    }
+    /* Prevent overwriting */
+    for(; j < w; j++) {
+        grad->map[j] = grad->map[j & 7];
+    }
+}
+
+
+#if LV_DITHER_ERROR_DIFFUSION == 1
+LV_ATTRIBUTE_FAST_MEM void lv_dither_fs_hor(lv_gradient_cache_t * grad, lv_coord_t xs, lv_coord_t y, lv_coord_t w)
```

Comment:
What does `fs` mean in the function's name?

Reply:
fs is for the name of the inventors "Floyd" and "Steinberg". We can change to "ed" instead for error diffusion.

Reply:
`ed` seems more intuitive, but `err_diff` or so sounds even better to me.